samtranslator/model/sam_resources.py

```diff
@@ -2194,3 +2202,108 @@ def _construct_appsync_schema(self, api_id: Intrinsicable[str]) -> GraphQLSchema
+    #       The proposed connector change will come in a follow up PR.
+    resource_type = "AWS::Serverless::GraphQLDataSource"
+    property_types = {
+        "ApiId": Property(True, one_of(IS_STR, IS_DICT)),
+        "Name": Property(False, IS_STR),
+        "Type": Property(True, IS_STR),
+        "Description": PassThroughProperty(False),
+        "ServiceRoleArn": Property(False, IS_STR),
+        "Permissions": Property(False, list_of(IS_STR)),
+        "DataSourceConfig": Property(True, IS_DICT),
+    }
+
+    ApiId: Intrinsicable[str]
+    Type: str
+    DataSourceConfig: DataSourceConfig
+    Name: Optional[str]
+    Description: Optional[PassThrough]
+    ServiceRoleArn: Optional[str]
+
+    @cw_timer
+    def to_cloudformation(self, **kwargs: Any) -> List[Resource]:
+        appsync_datasource = self._construct_appsync_datasource()
+        resources: List[Resource] = [appsync_datasource]
+
+        return resources
+
+    def _validate_dynamodb_datasource(self, ddb_properties: Dict[str, Any]) -> DynamoDBConfig:
+        ddb_config = {}
+
+        ddb_config["TableName"] = sam_expect(
+            ddb_properties.get("TableName"),
+            self.logical_id,
+            "DataSourceConfig.DynamoDB.TableName",
+        ).to_be_a_string()
+
+        ddb_config["AwsRegion"] = ddb_properties["Region"] if "Region" in ddb_properties else ref("AWS::Region")
+
+        if "DeltaSync" in ddb_properties:
+            deltasync_properties = ddb_properties["DeltaSync"]
+
+            sam_expect(
+                deltasync_properties.get("BaseTableTTL"), self.logical_id, "DataSourceConfig.DeltaSync.BaseTableTTL"
+            ).to_be_a_string()
+
+            sam_expect(
+                deltasync_properties.get("DeltaSyncTableName"),
+                self.logical_id,
+                "DataSourceConfig.DeltaSync.DeltaSyncTableName",
+            ).to_be_a_string()
+
+            sam_expect(
+                deltasync_properties.get("DeltaSyncTableTTL"),
+                self.logical_id,
+                "DataSourceConfig.DeltaSync.DeltaSyncTableTTL",
+            ).to_be_a_string()
+
+            ddb_config["DeltaSyncConfig"] = deltasync_properties
+
+        if "UseCallerCredentials" in ddb_properties:
+            ddb_config["UseCallerCredentials"] = ddb_properties["UseCallerCredentials"]
+
+        if "Versioned" in ddb_properties:
+            ddb_config["Versioned"] = ddb_properties["Versioned"]
+
+        return cast(DynamoDBConfig, ddb_config)
+
+    def _validate_config_properties(self, datasource: DataSource) -> None:
+        # DataSourceConfig is quite large property so we require a lot of additional validation.
+        # The datasource object is modified by reference.
+
+        if len(self.DataSourceConfig) > 1:
+            raise InvalidResourceException(
+                self.logical_id, "Only one resource can be defined at a time in 'DataSourceConfig'."
+            )
+
+        if datasource.Type == "AMAZON_DYNAMODB":
```

Comment:
Can we define a parameter for `AMAZON_DYNAMODB` and use it here and when we define `SUPPORTED_DATASOURCES` instead of hardcoding.

Reply:
This solution feels like a lot of constants to me for values that aren't being used frequently. For the most part, these types are only ever being used once. Why shouldn't we hardcode in that case?

Reply:
I agree that if it's only used once, there is no need to parameterize it. But I've already seen this value used twice in this PR and I suspect we're going to use `AMAZON_DYNAMODB` data source type in the future PRs as well..

Reply:
Fair point. I think I'll leave it as is for now and update in future PR if there's frequent reuse happening.